crates/rust-analyzer/src/target_spec.rs

```diff
@@ -1,35 +1,164 @@
+            Some(TargetForCrateRoot::Cargo(cargo_ws, target)) => {
+                let target_data = &cargo_ws[target];
+                let package_data = &cargo_ws[target_data.package];
+                let res = CargoTargetSpec {
+                    workspace_root: cargo_ws.workspace_root().to_path_buf(),
+                    project_manifest: package_data.manifest.clone(),
+                    crate_id,
+                    package: cargo_ws.package_flag(package_data),
+                    target: target_data.name.clone(),
+                    target_kind: target_data.kind,
+                    required_features: target_data.required_features.clone(),
+                    features: package_data.features.keys().cloned().collect(),
+                };
+                Ok(Some(TargetSpec::Cargo(res)))
+            }
+            Some(TargetForCrateRoot::JsonProject(_, krate)) => match krate.target_spec {
+                Some(target) => {
+                    let manifest = ManifestPath::try_from(target.manifest_file).unwrap();
+                    let res = ProjectJsonSpec {
+                        workspace_root: manifest.parent().to_path_buf(),
+                        project_manifest: manifest,
+                        target_kind: target.target_kind,
+                        target_label: target.target_label,
+                        runnables: target.runnables,
+                    };
+                    Ok(Some(TargetSpec::ProjectJson(res)))
+                }
+                None => {
+                    tracing::debug!(?krate, "no target spec");
+                    return Ok(None);
+                }
+            },
+            None => {
+                tracing::debug!(?crate_id, "no target found");
+                return Ok(None);
+            }
+        }
+    }
+
+    pub(crate) fn project_manifest(&self) -> &AbsPath {
+        match self {
+            TargetSpec::Cargo(cargo) => &cargo.project_manifest,
+            TargetSpec::ProjectJson(project_json) => &project_json.project_manifest,
+        }
+    }
+
+    pub(crate) fn workspace_root(&self) -> &AbsPath {
+        match self {
+            TargetSpec::Cargo(cargo) => &cargo.workspace_root,
+            TargetSpec::ProjectJson(project_json) => &project_json.workspace_root,
+        }
+    }
+
+    pub(crate) fn target_kind(&self) -> TargetKind {
+        match self {
+            TargetSpec::Cargo(cargo) => cargo.target_kind,
+            TargetSpec::ProjectJson(project_json) => project_json.target_kind,
+        }
+    }
+}
+
 #[derive(Clone)]
 pub(crate) struct CargoTargetSpec {
     pub(crate) workspace_root: AbsPathBuf,
-    pub(crate) cargo_toml: ManifestPath,
+    /// [`project_manifest`] corresponds to the file defining a crate. With Cargo,
+    /// this will be a Cargo.toml, but with a non-Cargo build system, this might be
+    /// a `TARGETS` or `BUCK` file. Multiple, distinct crates can share a
+    /// single `project_manifest`.
```

Comment:
This comment confuses me given where its placed, this is the field of a `CargoTargetSpec`, so this should always be a `Cargo.toml`.

Reply:
Ah—the intention behind this comment is to communicate that a non-Cargo project can also have a `TargetSpec` (née `CargoTargetSpec`), with the intention of trying to make non-Cargo build systems be more Cargo-like in order to reduce the maintenance burden on `rust-analyzer`, but as mentioned elsewhere, that might be a design bug.

Reply:
I get that, it's just that the location for this command seems ill placed, though i don't have a better spot for it either aside from somewhere in the module documentation.